design: rgw: add tenant and defaultPlacement fields to CephObjectStoreUser (#17755)
rook/rook: design/ceph/rgw-user-multitenancy-and-placement.md

# RGW User Multitenancy and Default Placement Targeting in CephObjectStoreUser

- **Issue**: https://github.com/rook/rook/issues/17274

## Summary

This document proposes extending the `CephObjectStoreUser` CRD with two new spec fields:

- `tenant` — assigns the RGW user to a named RGW tenant, enabling bucket name isolation across tenants.

> - `defaultPlacement` — sets the user's default bucket placement target, controlling which data/metadata pools newly created buckets land in.

Reviewer: Can I specify `defaultPlacement` without specifying `tenant`? I believe that's the case but want to make sure I understand fully.

Reviewer 2: RGW allows placement without tenant.  I have manually configured users setup that way. cephosu should support that as well.

Author: yes, it would simply that user is associated with that placement, all the data associated with this user will land to the placement we defined, with tenant coming into picture, it would mean multiple tenants can share the pool / namespace without collisions as they will be tagged with tenant prefix kindof.

Both fields already exist in the underlying `admin.User` struct in `go-ceph`; this work wires them into the Rook controller and API.

## Motivation

### Tenant Isolation

Ceph RGW supports a multitenancy model where users live in named tenants. Users in different tenants can own buckets with the same name without collision:

```
# Two separate objects, no conflict
tenantA$user1 → s3://photos

> tenantB$user1 → s3://photos

Reviewer: What does the awscli S3 config look like for tenantA and tenantB? Do they need to address the bucket with any special config modifications? I don't think this changes the design, but it may change what we document for users.

Author: Nope, they will just get the associated access and secret key with scaope of the placement they are associated with, this might be something worth documenting atleast here, I updated a section for it.

```

Rook currently has no mechanism to place a `CephObjectStoreUser` in an RGW tenant. Operators who need per-tenant user isolation must manage RGW users manually outside of Rook, forgoing the benefits of the operator (secret rotation, lifecycle management, status conditions).

### Placement Targeting

`CephObjectStore.spec.sharedPools.poolPlacements` already allows defining named placement targets (each backed by distinct metadata/data pools). However, the `CephObjectStoreUser` controller has no way to assign a user's `default-placement`, meaning all users default to the store-wide default placement. For multi-tier storage scenarios (hot/cold pools, per-tenant pools), operators need per-user placement control.

## Goals

> - Add `spec.tenant` to `CephObjectStoreUser` to pass `--tenant` to radosgw-admin at user creation.

Reviewer: Nit: Rook doesn't use radosgw-admin to provision COSUser. Admin Ops API instead.

    Tenant string `json:"tenant,omitempty"`

    // DefaultPlacement names the placement target for new buckets created by this user.
    // Must match one of the entries in the referenced CephObjectStore's
    // spec.sharedPools.poolPlacements[].name.
    // +optional
    // +kubebuilder:validation:Pattern=`^[a-zA-Z0-9._/-]+$`
    // +kubebuilder:validation:MaxLength=255
    DefaultPlacement string `json:"defaultPlacement,omitempty"`
}
```

### Example CR

```yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStoreUser
metadata:
  name: user1
  namespace: rook-ceph
spec:
  store: my-store
  displayName: "Tenant A User 1"
  tenant: tenantA
  defaultPlacement: hot-tier
```

## Immutability

`tenant` is immutable because RGW does not support moving a user between tenants; the only path is deletion and recreation. Attempting to change `tenant` on an existing `CephObjectStoreUser` would silently create a second user in the new tenant while leaving the original orphaned.

**Enforcement mechanism**: CEL `XValidation` rule on the spec field (same pattern as `accountRef`):
```go
// +kubebuilder:validation:XValidation:message="tenant is immutable",rule="self == oldSelf"
```

> This gives a clear admission webhook error on any attempted change.

Reviewer: ```suggestion
This gives a clear error on any attempted change.
```

## Proposed API Changes

### `ObjectStoreUserSpec` (`pkg/apis/ceph.rook.io/v1/types.go`)

```go
type ObjectStoreUserSpec struct {
    Store        string `json:"store,omitempty"`
    DisplayName  string `json:"displayName,omitempty"`
    // ... existing fields ...

    // Tenant is the RGW tenant this user belongs to.
    // Users in different tenants can have buckets with the same name without conflict.
    // When set, the effective user ID in RGW becomes "<tenant>$<name>".
    // This field is immutable after creation.
    // +optional
    // +kubebuilder:validation:XValidation:message="tenant is immutable",rule="self == oldSelf"
    // +kubebuilder:validation:Pattern=`^[a-zA-Z0-9._-]+$`
    // +kubebuilder:validation:MaxLength=255
    Tenant string `json:"tenant,omitempty"`

    // Placement controls the user's default bucket placement target and storage class tags.
    // May be set independently of tenant.
    // +optional
    // +nullable

>     Placement *ObjectStoreUserPlacementSpec `json:"placement,omitempty"`

Reviewer 2: What is the expected behavior if `.placement` is set and then removed? There have been several discussion lately what the correct behavior is when a field is unset -- this is something that should probably be part of the design blueprint.

Reviewer 2: I think it probably an aesthetic decision if there is a `ObjectStoreUserPlacementSpec` or if the 3 placement related fields are directly under `ObjectStoreUserSpec`.  I favor being directly under `ObjectStoreUserSpec` since that mirrors the format of go-ceph's User but I will differ to others judgement on what makes a better interface.

Author: makes sense @jhoblitt i updated the design doc, can you take another look. thanks!

}

// ObjectStoreUserPlacementSpec sets the user's default placement target and storage class tags.
type ObjectStoreUserPlacementSpec struct {

>     // ID names the placement target for new buckets created by this user.

Reviewer 2: `User.DefaultStorageClass` should be mentioned in this doc since it is so closely tied to `User.DefaultPlacement `.

The rgw/go-ceph handling of the default storage class is kind of strange. It has to be set by via the placement field as `<placement>/<storage-class>`. Then then the `<storage-class>` value comes out of `User.DefaultStorageClass`.  This is what was required to get this working: https://github.com/rook/rook/pull/17260/changes#diff-4399505e3e48def2feaa495cba5133401361caddd0f494399172bab3acb15a64R660

I think it should be explicitly called out if setting  `<placement>/<storage-class>` via the "ID" field is supported or if it is disallowed and setting "ID" and `DefaultStorageClass` together is required to achieved that effect.

    // Must match one of the entries in the referenced CephObjectStore's
    // spec.sharedPools.poolPlacements[].name.
    // +optional
    // +kubebuilder:validation:Pattern=`^[a-zA-Z0-9._/-]+$`
    // +kubebuilder:validation:MaxLength=255

>     ID string `json:"id,omitempty"`

Reviewer 2: I dislike using `ID` as a field name. It doesn't map to a single `User` field that's related to placement (`DefaultPlacement`, `DefaultStorageClass`, and `PlacementTags` all have IDs). It also doesn't map well to the ceph cli tools which use "key".  This field may also need to support setting both `DefaultPlacement` and `DefaultStorageClass`.

https://github.com/ceph/go-ceph/blob/master/rgw/admin/user.go#L22-L23

```
# Two separate objects, no conflict
tenantA$user1 → s3://photos
tenantB$user1 → s3://photos
```

Rook currently has no mechanism to place a `CephObjectStoreUser` in an RGW
tenant. Operators who need per-tenant user isolation must manage RGW users
manually outside of Rook, forgoing the benefits of the operator (secret
rotation, lifecycle management, status conditions).

### Placement Targeting

`CephObjectStore.spec.sharedPools.poolPlacements` already allows defining named
placement targets (each backed by distinct metadata/data pools). However, the
`CephObjectStoreUser` controller has no way to assign a user's
`default-placement`, meaning all users default to the store-wide default
placement.

## Goals

- Add `spec.tenant` to `CephObjectStoreUser` to set the `Tenant` field on the RGW Admin Ops API `User` struct at user creation.
- Add `spec.defaultPlacement` to `CephObjectStoreUser` to set `DefaultPlacement` on the user via `CreateUser`/`ModifyUser` in the RGW Admin Ops API.
- Add `spec.defaultStorageClass` to `CephObjectStoreUser` to set `DefaultStorageClass` on the user, embedded into the placement rule sent to RGW

> - Add `spec.defaultPlacementTags` to `CephObjectStoreUser` to set `PlacementTags` on the user via `CreateUser`/`ModifyUser`, now that go-ceph supports it upstream (merged in [ceph/go-ceph#1290](https://github.com/ceph/go-ceph/pull/1290)).

Reviewer: I considered whether it might be helpful for Rook users to have these under a `spec.placement` parent struct. After consideration, I think the flat hierarchy here is sufficient and less complex, thus preferable.

I thought I would mention it and ask for @travisn and @jhoblitt input as well, since I am still slightly on the fence.

Reviewer: This is the main source of my still being on the fence

> Require `defaultPlacement` to be set whenever `defaultStorageClass` is set, since RGW cannot apply a storage class without a placement target.

Author: Agreed on keeping it flat: I've written the reasoning into the doc so it's on record rather than just living in this thread.

With flat hierarchy, It matches `go-ceph`'s `admin.User`, which is what the controller marshals into.

Reviewer: We shouldn't always duplicate Ceph or go-ceph's APIs. That kind of pattern tends to create leaky abstractions over time. In this case, I think it's appropriate, but I do want to make sure that it's clear that our focus should be on what the best API is for users and not what is easiest for developers.

Author: Agreed, its a difficult decision to choose flat when there is some dependence on a field.

from a user perspective: these are three things you set on a user, and two are independently useful (`defaultPlacement` and `defaultPlacementTags` each work alone; only `defaultStorageClass` has a prerequisite). If we use `spec.placement` 
A `spec.placement` struct would present them as one composite you configure together, which describes actual usage less accurately than three similar configs with one documented dependency.

Where copying upstream would confuse users, the design already avoids it : users get separate `defaultPlacement` and `defaultStorageClass` fields instead of RGW's packed `<placement>/<storage-class>` string, a separate `tenant` field instead of the `tenant$uid` form, and `defaultPlacementTags` rather than go-ceph's `PlacementTags`.

Happy to go nested if you or @travisn think that nesting just the storageclass and placement can be useful

manually outside of Rook, forgoing the benefits of the operator (secret
rotation, lifecycle management, status conditions).

### Placement Targeting

`CephObjectStore.spec.sharedPools.poolPlacements` already allows defining named
placement targets (each backed by distinct metadata/data pools). However, the
`CephObjectStoreUser` controller has no way to assign a user's
`default-placement`, meaning all users default to the store-wide default
placement.

## Goals

- Add `spec.tenant` to `CephObjectStoreUser` to set the `Tenant` field on the RGW Admin Ops API `User` struct at user creation.
- Add `spec.defaultPlacement` to `CephObjectStoreUser` to set `DefaultPlacement` on the user via `CreateUser`/`ModifyUser` in the RGW Admin Ops API.
- Add `spec.defaultStorageClass` to `CephObjectStoreUser` to set `DefaultStorageClass` on the user, embedded into the placement rule sent to RGW
- Add `spec.defaultPlacementTags` to `CephObjectStoreUser` to set `PlacementTags` on the user via `CreateUser`/`ModifyUser`, now that go-ceph supports it upstream (merged in [ceph/go-ceph#1290](https://github.com/ceph/go-ceph/pull/1290)).

> - Validate `defaultPlacement` against the named placements defined in the referenced `CephObjectStore`.

Reviewer 2: > This is @jhoblitt's AI agent.

Matching Goals-level change for the validation-source correction on the field godoc: operator-side validation against the store CR is wrong for zone-backed stores and impossible for external ones, while RGW already validates every write against the live zonegroup.

```suggestion
- Leave `defaultPlacement` validation to RGW: the serving RGW validates it
  against the live zonegroup on every create/modify and rejects unknown
  targets with `EINVAL`; the operator surfaces that failure in the CR status.
  (Validating against the referenced `CephObjectStore`'s `spec.sharedPools`
  would be wrong for zone-backed stores, whose placements live on
  `CephObjectZone`, and impossible for external stores.)
```

— the Overlord

- All four fields are independent of `tenant`: `defaultPlacement`/`defaultStorageClass`/`defaultPlacementTags` may be set without `tenant`, and vice versa.
- Preserve backward compatibility: all new fields are optional, existing resources are unaffected.

## Background

### RGW Tenant User ID Format

When a user is created in a tenant, RGW stores and returns the user as `$tenant$uid` internally, but the external UID (passed to the API) is just `uid`. When querying or deleting a tenanted user, the RGW Admin Ops API accepts a `Tenant` field alongside `ID` rather than a combined string.

Equivalently via CLI:
```bash
radosgw-admin user create --uid="user1" --tenant="tenantA" --display-name="User 1"
# effective UID internally: tenantA$user1

radosgw-admin user info --uid="user1" --tenant="tenantA"

> ```

Reviewer 2: > This is @jhoblitt's AI agent.

This sentence describes an API contract RGW does not provide, and it matters because the implementation PR (#17792) copied it verbatim. RGW parses a separate `tenant` request parameter **only in user create** — the sole `tenant` RESTArg in `rgw_rest_user.cc` on squid, tentacle, and main; user info/modify/remove resolve tenancy solely through the combined `uid` (`rgw_user::from_str` splits on `$`). go-ceph matches: `tenant` is in `CreateUser`'s URL-parameter whitelist only, while `GetUser` sends `{uid, access-key, stats}`, `RemoveUser` `{uid, purge-data}`, and `ModifyUser` has no tenant — true in released go-ceph, on master, and in the fork #17792 pins.

Built as written, every non-create operation therefore addresses the bare `uid` — the same-named user in the *default* tenant. If an untenanted namesake exists (and the feature's headline use case, same short uid across tenants, is what makes those likely), reconcile silently adopts it: its keys are published into the tenanted CR's Secret, its other keys are purged, its caps/quota get modified, and CR deletion removes *it* — while the tenanted user is never created; every call returns 200 and the CR reads Ready. With no namesake, the first reconcile converges by accident, the second wedges in a CreateUser → `UserAlreadyExists` failure loop, and CR deletion orphans `tenantA$user1` with live credentials.

The fix needs no upstream changes — the combined-uid form is uniformly supported by every op on every branch:

````suggestion
When a user is created in a tenant, the user's identity everywhere in RGW is the combined form `<tenant>$<uid>`. Every Admin Ops operation resolves the `uid` parameter through this form (`rgw_user::from_str` splits on `$`); a bare `uid` addresses the user in the default (empty) tenant.

The Admin Ops API accepts a separate `tenant` parameter **only on user create**. User info, modify, and remove have no tenant parameter — on those operations a tenant can only be expressed inside the combined `uid`. go-ceph mirrors this: `admin.User.Tenant` is transmitted by `CreateUser` only and silently dropped by `GetUser`/`ModifyUser`/`RemoveUser`, whose URL-parameter whitelists exclude `tenant`.

The controller therefore uses the combined `<tenant>$<name>` string as the user ID for **every** Admin Ops call, create included, and never relies on the go-ceph `Tenant` struct field. Addressing a tenanted user by bare `uid` would silently resolve to the same-named user in the default tenant: reconciliation would adopt (and CR deletion would delete) an unrelated user, or wedge in a CreateUser/UserAlreadyExists loop when none exists. As a backstop, the reconcile fails with an explicit error if the live user's tenant does not match `spec.tenant`, rather than adopting a user from another tenant.

Equivalently via CLI:
```bash
radosgw-admin user create --uid="tenantA$user1" --display-name="User 1"

radosgw-admin user info --uid="tenantA$user1"
```
````

— the Overlord

    Tenant              string        `url:"tenant"`
    DefaultPlacement    string        `json:"default_placement" url:"default-placement"`
    DefaultStorageClass string        `json:"default_storage_class"`   // ← no url tag: not sent
    PlacementTags       []interface{} `json:"placement_tags"`          // ← no url tag: not sent
    // ...
}

// what this design requires (ceph/go-ceph#1290, pending a tagged release)
type User struct {
    ID                  string   `json:"user_id" url:"uid"`
    Tenant              string   `url:"tenant"`                        // ← URL param only, not in JSON response
    DefaultPlacement    string   `json:"default_placement" url:"default-placement"`
    DefaultStorageClass string   `json:"default_storage_class" url:"default-storage-class"`
    PlacementTags       []string `json:"placement_tags" url:"placement-tags"`
    // ...
}
```

This gap is why `spec.defaultStorageClass` and `spec.defaultPlacementTags`
cannot be implemented against a released `go-ceph` today, and why `go.mod`
currently points at a fork. See [Setting the storage
class](#setting-the-storage-class) for why it also rules out the workaround of

> packing the storage class into the placement string.

Reviewer 2: > This is @jhoblitt's AI agent.

This block is stale in three ways: Rook's `go.mod` pins go-ceph **v0.40.0** (not v0.39.0), which already transmits `default-placement` *and* `default-storage-class` on both create and modify; the repo does not point at a fork (that pin existed only on an implementation branch); and the awaited release exists — **v0.41.0** shipped 2026-08-11 with placement-tags support. The dependency story reduces to a routine `go.mod` bump.

````suggestion
The `go-ceph` `admin.User` struct models all four fields. Released support
arrived in two steps: v0.40.0 (Rook's current pin) transmits
`default-placement` and `default-storage-class` on both `CreateUser` and
`ModifyUser`; v0.41.0 (released 2026-08-11) adds `placement-tags`
([ceph/go-ceph#1290](https://github.com/ceph/go-ceph/pull/1290)):

```go
// go-ceph v0.41.0 (released)
type User struct {
    ID                  string   `json:"user_id" url:"uid"`
    Tenant              string   `url:"tenant"`  // ← URL param on user create only
    DefaultPlacement    string   `json:"default_placement" url:"default-placement"`
    DefaultStorageClass string   `json:"default_storage_class" url:"default-storage-class"`
    PlacementTags       []string `json:"placement_tags" url:"placement-tags"`
    // ...
}
```

Implementing `spec.defaultPlacementTags` therefore requires only a `go.mod`
bump to go-ceph v0.41.0 — no fork and no unreleased dependency. See
[Setting the storage class](#setting-the-storage-class) for how the two
placement parameters are applied on the wire.
````

— the Overlord

Author: added short note in future work section

// +kubebuilder:validation:XValidation:message="defaultPlacement cannot be unset once set; RGW does not support clearing a user's default placement",rule="!has(oldSelf.defaultPlacement) || has(self.defaultPlacement)"
// +kubebuilder:validation:XValidation:message="defaultStorageClass cannot be unset once set; RGW does not support clearing a user's default storage class",rule="!has(oldSelf.defaultStorageClass) || has(self.defaultStorageClass)"
// +kubebuilder:validation:XValidation:message="defaultPlacementTags cannot be unset once set; RGW does not support clearing a user's placement tags",rule="!has(oldSelf.defaultPlacementTags) || has(self.defaultPlacementTags)"
type ObjectStoreUserSpec struct {
    Store        string `json:"store,omitempty"`
    DisplayName  string `json:"displayName,omitempty"`
    // ... existing fields ...

    // Tenant is the RGW tenant this user belongs to.
    // Users in different tenants can have buckets with the same name without conflict.
    // When set, the effective user ID in RGW becomes "<tenant>$<name>".
    // This field is immutable after creation.
    // +optional

>     // +kubebuilder:validation:XValidation:message="tenant is immutable",rule="self == oldSelf"

Reviewer 2: > This is @jhoblitt's AI agent.

The field-level `self == oldSelf` rule doesn't deliver the immutability promised here: Kubernetes skips a transition rule on an **optional** field when an update sets or unsets that field ([CRD validation docs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#transition-rules): "When placed on an optional field, a transition rule will not apply to update operations that set or unset the field"). So `tenant` can still be **added to** an existing tenantless user, or **removed from** a tenanted one — exactly the two transitions whose consequence the Immutability section describes: on the next reconcile the controller creates a second RGW user under the other identity, adopts its keys, and silently repoints the CR's Secret, orphaning the original user and its buckets.

The spec-level form below closes both paths. It follows the same pattern as the three unset rules already above it (spec-level transition rules always evaluate, because the spec object exists on both sides of every update), stays legal on create (transition rules don't run there, so setting `tenant` at creation still works), and invalidates no stored object (absent-to-absent passes). The equivalent alternative is keeping the field-level rule with `optionalOldSelf=true` (k8s >= 1.29; Rook's floor is v1.31).

Because the CEL rule lives in the CRD, a cluster running a not-yet-upgraded CRD still admits the transition — worth pairing this with a controller-side backstop: fail the reconcile with an explicit error when the live user's tenant differs from `spec.tenant`, rather than adopting or creating.

```suggestion
// +kubebuilder:validation:XValidation:message="defaultStorageClass requires defaultPlacement",rule="!has(self.defaultStorageClass) || has(self.defaultPlacement)"
// +kubebuilder:validation:XValidation:message="defaultPlacement cannot be unset once set; RGW does not support clearing a user's default placement",rule="!has(oldSelf.defaultPlacement) || has(self.defaultPlacement)"
// +kubebuilder:validation:XValidation:message="defaultStorageClass cannot be unset once set; RGW does not support clearing a user's default storage class",rule="!has(oldSelf.defaultStorageClass) || has(self.defaultStorageClass)"
// +kubebuilder:validation:XValidation:message="defaultPlacementTags cannot be unset once set; RGW does not support clearing a user's placement tags",rule="!has(oldSelf.defaultPlacementTags) || has(self.defaultPlacementTags)"
// +kubebuilder:validation:XValidation:message="tenant is immutable",rule="has(oldSelf.tenant) == has(self.tenant) && (!has(self.tenant) || self.tenant == oldSelf.tenant)"
type ObjectStoreUserSpec struct {
    Store        string `json:"store,omitempty"`
    DisplayName  string `json:"displayName,omitempty"`
    // ... existing fields ...

    // Tenant is the RGW tenant this user belongs to.
    // Users in different tenants can have buckets with the same name without conflict.
    // When set, the effective user ID in RGW becomes "<tenant>$<name>".
    // This field is immutable after creation.
    // +optional
```

— the Overlord

Author: fixed this

>     // +kubebuilder:validation:Pattern=`^[a-zA-Z0-9._-]+$`

Reviewer 2: > This is @jhoblitt's AI agent.

RGW's tenant validation (`rgw_validate_tenant_name`, `src/rgw/rgw_user.cc` — enforced on the admin-ops create/modify path) accepts only alphanumerics and `_`, rejecting anything else with `ERR_INVALID_TENANT_NAME`. The current pattern admits `.` and `-`, so `tenant: team-a` passes admission and then permanently fails reconcile on an immutable field. (`MaxLength=255` is fine to keep as a Rook-side bound — RGW imposes no length limit, so it just shouldn't be read as RGW-derived.)

```suggestion
    // +kubebuilder:validation:Pattern=`^[a-zA-Z0-9_]+$`
```

— the Overlord

    // +kubebuilder:validation:MaxLength=255
    Tenant string `json:"tenant,omitempty"`

    // DefaultPlacement overrides the default pool placement for buckets created by
    // this user. Must match one of the entries in the referenced CephObjectStore's
    // spec.sharedPools.poolPlacements[].name. If not provided, the zone group's
    // default placement target is used.
    // This field cannot be unset once set — see "Unset/Removal Behavior".
    // +optional
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=2048

>     DefaultPlacement *string `json:"defaultPlacement,omitempty"`

Reviewer 2: > This is @jhoblitt's AI agent.

Two changes here. (1) The store-CR reference is the wrong validation source: zone-backed stores take their placements from `CephObjectZone.spec.sharedPools` (the store's own `sharedPools` is validated but never applied for multisite stores), and external-mode stores — which fully support `CephObjectStoreUser` — have no rook CR carrying a placement list at all. RGW's own `valid_placement` check against the live zonegroup is the only validator that is correct in every topology, and it already runs on every create/modify. (2) Adding the no-`/` pattern rejects the packed `placement/storage-class` form at admission and keeps the value's meaning identical across Ceph versions (`/` is the embed separator in the squid-era wire encoding).

```suggestion
    // DefaultPlacement overrides the default pool placement for buckets created by
    // this user. It must name a placement target known to the zonegroup serving
    // the referenced object store; RGW validates this on every create and modify
    // and rejects unknown targets. If not provided, the zone group's
    // default placement target is used.
    // This field cannot be unset once set — see "Unset/Removal Behavior".
    // +optional
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=2048
    // +kubebuilder:validation:Pattern=`^[a-zA-Z0-9._-]+$`
    DefaultPlacement *string `json:"defaultPlacement,omitempty"`
```

— the Overlord

Author: addressed in followup pr

    // DefaultStorageClass overrides the default storage class for objects created by
    // this user. Requires DefaultPlacement to be set. If not provided, the default
    // `STANDARD` storage class is used.
    // This field cannot be unset once set — see "Unset/Removal Behavior".
    // +optional
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=2048
    DefaultStorageClass *string `json:"defaultStorageClass,omitempty"`

    // DefaultPlacementTags is a list of storage class tags to associate with this

>     // user's default placement.

Reviewer 2: > This is @jhoblitt's AI agent.

RGW `placement_tags` is a bucket-creation authorization list, not storage-class metadata, and it is not scoped to the default placement: when a zonegroup placement target carries tags, only users holding a matching tag may create buckets there; untagged targets stay open to everyone (docs.ceph.com/en/latest/radosgw/placement/). The squashed commit message already describes this correctly — the doc text contradicts it. The Summary bullet and the Goals bullet for this field want the same rewording, and renaming to `placementTags` (dropping the `default` prefix, which wrongly implies default-placement scoping) is worth weighing while the name is still free.

```suggestion
    // DefaultPlacementTags restricts which placement targets this user may
    // create buckets in: if a zonegroup placement target carries tags, RGW
    // allows bucket creation there only when one of the user's placement
    // tags matches. Targets without tags remain usable by every user. The
    // tags are not storage-class related and are not scoped to DefaultPlacement.
```

— the Overlord

- `DefaultPlacement` → `DefaultPlacement` (URL param `default-placement`)
- `DefaultStorageClass` → `DefaultStorageClass` (URL param `default-storage-class`)
- `DefaultPlacementTags` → `PlacementTags` (URL param `placement-tags`, JSON `placement_tags`)

### Setting the storage class

RGW stores a user's default placement as a single `rgw_placement_rule` value
with two members, `name` and `storage_class` (`src/rgw/rgw_placement_types.h`),
but it does not present that value as one field on either side of the admin ops
API:

- **Reading**, the two members are split across two JSON keys —
  `encode_json("default_placement", default_placement.name, f)` and
  `encode_json("default_storage_class", default_placement.get_storage_class(), f)`
  in `RGWUserInfo::dump` (`src/rgw/rgw_common.cc`). Note `get_storage_class()`
  canonicalizes an unset storage class to `STANDARD`, so a user that has never
  had one set still reports `"default_storage_class": "STANDARD"`.
- **Writing**, they are two separate URL params, and RGW composes the rule
  itself: `target_rule.name = default_placement_str`, then
  `target_rule.storage_class = default_storage_class_str` if non-empty

>   (`RGWOp_User_Modify::execute`).

Reviewer 2: > This is @jhoblitt's AI agent.

The write path described here is Tentacle-and-later behavior, not RGW behavior generally: [ceph#57985](https://github.com/ceph/ceph/pull/57985) (merged 2024-07, fixes [tracker 66439](https://tracker.ceph.com/issues/66439)) was never backported, so on squid `default-storage-class` is never parsed and the handler builds the rule with `target_rule.from_str(default_placement_str)` — splitting on `/`. Verified against released tags: v19.2.3 has `from_str` and no `default-storage-class` parsing; v20.2.0 has the verbatim assignment plus the separate param. Rook currently supports both majors (`pkg/operator/ceph/version/version.go`: `supportedVersions = {Squid, Tentacle}`), so with a single unconditional encoding one of them always loses: separate params are **silently ignored on v19** (user asked for `STANDARD_IA`, gets `STANDARD`, no error, CR Ready), while the embedded form **fails with `EINVAL` on v20**. The read side has the same one-version generalization in the other direction: `get_storage_class()` canonicalization in `RGWUserInfo::dump` exists only on main — squid and tentacle dump the raw member, so an unset storage class reads as `""` on both supported majors (and will flip to `"STANDARD"` under a future Ceph upgrade).

Recommendation: keep this section's user-facing contract exactly as written — two spec fields, no packing in `spec.defaultPlacement` — and make the controller select the wire encoding internally by `cephver` (`generateUserConfig` already receives the Ceph version on master). The CR's behavior stays invariant across the support window, which we'd like to hold as a hard requirement: RGW admin-ops API differences must be absorbed by the controller, never surfaced as version-conditional CRD behavior. The squid arm is a temporary internal encoding tied to tracker 66439, deleted when squid ages out.

```suggestion
- **Reading**, the two members are split across two JSON keys —
  `"default_placement"` and `"default_storage_class"` in `RGWUserInfo::dump`
  (`src/rgw/rgw_common.cc`). On squid and tentacle the raw `storage_class`
  member is dumped, so a user that has never had one set reports
  `"default_storage_class": ""`; only on main (v21) does `get_storage_class()`
  canonicalize it to `STANDARD`. The controller treats `""` and `STANDARD`
  as equivalent when comparing desired and live state.
- **Writing**, the wire encoding differs by Ceph version, and the controller
  selects it internally by `cephver` so the CRD contract is identical on
  every supported release:
  - **Tentacle (v20) and later** ([ceph#57985](https://github.com/ceph/ceph/pull/57985),
    [tracker 66439](https://tracker.ceph.com/issues/66439), never backported
    to squid): two separate URL params, and RGW composes the rule itself —
    `target_rule.name = default_placement_str`, then
    `target_rule.storage_class = default_storage_class_str` if non-empty
    (`RGWOp_User_Modify::execute`).
  - **Squid (v19)**: `default-storage-class` is never parsed; the handler
    builds the rule with `target_rule.from_str(default_placement_str)`, which
    splits on the first `/`. On squid only, the controller composes
    `<placement>/<storage-class>` into the `default-placement` parameter — an
    internal, explicitly temporary encoding, removed when squid leaves Rook's
    support window.

  Both encodings produce the identical `rgw_placement_rule` on the user, are
  validated by the same server-side `valid_placement` check (an unknown
  placement or storage class fails with `EINVAL` on both), and are reported
  back identically by user info — so reconcile comparisons, status, and the
  CRD contract are version-blind. Unit tests assert the exact wire fields
  `generateUserConfig` produces for each version.
```

Two knock-on edits if this is taken: the paragraph below ("RGW assigns the parameter to `target_rule.name` verbatim without splitting on `/`") wants an "on Tentacle and later" qualifier, since squid's split is exactly what the internal squid arm relies on — the *users never see or construct the `/`-joined form* conclusion stays true either way; and the `no-/` restriction on `spec.defaultPlacement` becomes load-bearing for version invariance (a slash-free name means the same thing under both encodings), worth a `Pattern` on the field rather than prose alone.

— the Overlord

The `<placement>/<storage-class>` form that appears elsewhere in Ceph is the
*serialized* representation of that struct — `rgw_placement_rule::to_str()`
emits it for on-disk encoding, and `from_str()` splits it back apart on decode.
It is not the admin ops wire format. **Packing a storage class into
`spec.defaultPlacement` (e.g. `hot-tier/STANDARD_IA`) is therefore not
supported, and this design disallows it.** RGW assigns the parameter to
`target_rule.name` verbatim without splitting on `/`, and then validates it with
`RGWZoneParams::valid_placement`, which does a `placement_pools.find(rule.name)`
lookup (`src/rgw/rgw_zone.h`). A packed string will not match any configured
placement target, so the request is rejected with `EINVAL`. Because
`spec.defaultPlacement` is separately validated against the referenced
`CephObjectStore`'s `spec.sharedPools.poolPlacements[].name`, a packed value is

> rejected by Rook before it ever reaches RGW.

Reviewer 2: > This is @jhoblitt's AI agent.

With validation moved to RGW (see the Goals/godoc suggestions), the operator-side check this sentence leans on no longer exists — the no-`/` pattern on the field is what rejects packed values at admission.

```suggestion
lookup (`src/rgw/rgw_zone.h`). A packed string will not match any configured
placement target, so the request is rejected with `EINVAL`. The no-`/`
pattern on `spec.defaultPlacement` additionally rejects a packed value at
admission, before it ever reaches RGW.
```

— the Overlord

So the two Rook fields map one-to-one onto the two URL params, and the operator
sends them separately. Users never see or construct the `/`-joined form. The
reason this needed clarifying is historical: because released `go-ceph` cannot
send `default-storage-class` at all (see [Background](#rgw-tenant-user-id-format)),
packing looked like the only available route to set a storage class — as hit in
[#17260](https://github.com/rook/rook/pull/17260). Fixing it in `go-ceph` is the
correct layer, which is what [ceph/go-ceph#1290](https://github.com/ceph/go-ceph/pull/1290)
does; hiding the joined string from Rook's API is deliberate, since it is an
RGW-internal encoding detail rather than something users should have to know.

> **`PlacementTags` was previously out of scope** because go-ceph did not yet support it and PR #17260 does not implement it. That gap has since closed: [ceph/go-ceph#1290](https://github.com/ceph/go-ceph/pull/1290) added `PlacementTags` support to the admin ops client and merged upstream on 2026-07-09. This design now includes it as a flat `DefaultPlacementTags []string` field, following the same flat, go-ceph-aligned naming as the other fields. Rook's `go.mod` currently points at a fork (`github.com/ideepika/go-ceph`) pending an official tagged go-ceph release that includes this change; that pin must be replaced with a released version before this can merge.

Reviewer 2: > This is @jhoblitt's AI agent.

Same currency fix as the Background block (v0.41.0 is released; no fork), plus the scoping caveat this field needs on record: tags only act against zonegroup placement targets that themselves carry tags, and Rook's `PoolPlacementSpec` has no `tags` field today — the enabling half of the mechanism.

```suggestion
**`PlacementTags` availability and scope:** [ceph/go-ceph#1290](https://github.com/ceph/go-ceph/pull/1290) added placement-tags support to the admin ops client and is included in released go-ceph v0.41.0, so this field only needs a `go.mod` bump — no fork. One caveat belongs on record: placement tags act only against zonegroup placement targets that themselves carry tags, and Rook's `PoolPlacementSpec` (on both `CephObjectStore` and `CephObjectZone`) has no `tags` field today. In a fully Rook-managed topology this knob therefore has no effect until targets are tagged out-of-band (`radosgw-admin zonegroup placement modify --tags ...`, which Rook's zonegroup reconcile preserves) or a `PoolPlacementSpec.tags` counterpart lands as follow-up work; deployments with externally managed zonegroups can use it as-is.
```

— the Overlord

Author: keeping thiis seperately in future work pr

with the CEL transition rules shown above, so the failure is a clear, immediate
validation error instead of drift the user cannot see. `MinLength=1` on the
string fields closes the obvious workaround of setting the field to `""`, which
would pass the CEL check and then be ignored by RGW. Users who need to return a
user to the zone group's default placement must delete and recreate the
`CephObjectStoreUser`. This caveat must be stated in the user-facing CRD
documentation as well as in the field's Go doc comment. When the upstream fixes
are released, the CEL rules can be relaxed and the controller taught to send the
explicit clear, moving these fields to the unset-means-disabled behavior we
want; because that change only removes a restriction, it is backward compatible.
Whether to additionally mark these fields experimental for the first release is
left to code/doc review.

### Example CR

```yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStoreUser
metadata:
  name: user1
  namespace: rook-ceph
spec:
  store: my-store
  displayName: "Tenant A User 1"
  tenant: tenantA
  defaultPlacement: hot-tier
  defaultStorageClass: STANDARD_IA
  defaultPlacementTags:
    - tenant-a
```

> ## S3 Client Configuration for Tenanted Users

Reviewer 2: > This is @jhoblitt's AI agent.

Two missing sections in one insertion. **Status**: this feature's characteristic failures are otherwise silent (an RGW-rejected placement, a tenant mismatch), and the Motivation already promises status reporting — this defines the minimal surface. **Multisite**: user placement fields are realm-synced metadata referencing zonegroup-scoped targets, which is a combination no other user field has; the doc should state the consequences (or scope multisite out explicitly). Trim as needed.

```suggestion
## Status

The controller echoes applied state into the CR status after a successful
reconcile: the effective `default_placement` and `default_storage_class`
read back from user info (current Ceph releases report an unset storage
class as `""`; the controller treats `""` and `STANDARD` as equivalent).
A placement or storage class rejected by RGW's server-side validation
fails the reconcile and surfaces the RGW error in the CR status; this is
the intended validation UX. A tenant mismatch between spec and the live
user is likewise a surfaced reconcile error, never a silent adoption.

## Multisite

RGW user metadata — including `default_placement` and
`default_storage_class` — is realm-scoped and replicates to every zone via
metadata sync. Placement *targets*, however, are zonegroup-scoped, and
their pools are zone-local. Consequences this design accepts:

- Validation happens at apply time, by the RGW serving the referenced
  object store, against **its** zonegroup only.
- In a realm with multiple zonegroups (or independently managed zone
  specs), a user's synced `default_placement` may name a target that does
  not exist in a peer zonegroup. Bucket creation there fails with
  `InvalidLocationConstraint` at the S3 layer; Rook does not detect this.
  Deployments using per-user placement across zonegroups should define the
  same placement target names in every zonegroup of the realm.
- Rook does not re-validate user placements when zonegroup placement
  targets change after the fact.

Zone-backed stores (`spec.zone.name` set) and external-mode stores are
fully supported: because validation is RGW-side, no rook CR needs to carry
the placement list.

## S3 Client Configuration for Tenanted Users
```

— the Overlord

Author: added this section

Users within the same tenant access their buckets using standard S3 virtual-host-style URLs with no changes:

```
my-bucket.s3.ceph.io   ← works normally for same-tenant users
```

RGW resolves the bucket to the correct tenant namespace based on the credentials used. No DNS changes or special endpoint configuration are required for this feature's primary use case.

### Cross-tenant access (out of scope, deprecated upstream)

Cross-tenant bucket access via path-style requests using the `tenant:bucket`
notation (e.g. `s3.ceph.io/tenantA:my-bucket/`) is a Ceph extension to the S3
protocol. As noted in the Ceph Tentacle release notes, this feature is
deprecated and scheduled for removal.

> S3 API support for cross-tenant names such as `Bucket='tenant:bucketname'`

Virtual-host-style cross-tenant access (`tenantA:my-bucket.s3.ceph.io`) is not
possible because `:` is not valid in DNS names.

**Cross-tenant bucket sharing is explicitly out of scope for this feature.**
Users who need to share buckets across tenant boundaries should be placed in
the same tenant namespace. This aligns with Ceph's upstream direction of
removing cross-tenant path-style access.

> ## Immutability

Reviewer 2: > This is @jhoblitt's AI agent.

`tenant` is the first `ObjectStoreUserSpec` field whose presence makes a one-minor operator rollback actively destructive rather than merely stale — the older operator addresses the user by bare name, misses the tenanted user, creates an untenanted namesake, and repoints the CR's Secret at it. That deserves a stated procedure and a release note.

```suggestion
## Compatibility and rollback

All fields are optional; CRs created by older Rook are unaffected, and the
new schema invalidates no stored object.

Rolling back to a Rook release that predates `spec.tenant` while tenanted
CRs exist is **destructive**: the older operator addresses the user by
bare name, fails to find the tenanted user, creates an untenanted user
with the same name, and repoints the CR's Secret at it — orphaning the
tenanted user and its buckets. Before downgrading, tenanted
`CephObjectStoreUser` CRs must be removed (or the operator scaled down).
This warning ships in the release notes. The placement fields carry no
such hazard: an older operator simply leaves the live values in place.

## Immutability
```

— the Overlord

Author: noted

`tenant` is immutable because RGW does not support moving a user between
tenants; the only path is deletion and recreation. Attempting to change
`tenant` on an existing `CephObjectStoreUser` would silently create a second
user in the new tenant while leaving the original orphaned.

`defaultPlacement`, `defaultStorageClass`, and `defaultPlacementTags` are
mutable — RGW supports changing a user's default placement, storage class, and
placement tags at any time; changes only affect future bucket/object creation,
not existing buckets/objects. Removal is the exception: as described in
[Unset/Removal Behavior](#unsetremoval-behavior), RGW offers no way to clear
these values, so the CRD rejects an update that unsets a previously set field.
This is a caveat of the initial implementation, not the intended end state.

## Interaction with `AccountRef`

`tenant` and `accountRef` are orthogonal. Both can be set simultaneously. When both are set:
- `userConfig.Tenant` is set from `spec.tenant`
- `userConfig.AccountID` is set from `spec.accountRef`

> No conflict; `admin.User` has both fields.

Reviewer 2: > This is @jhoblitt's AI agent.

"No conflict" holds for RGW in the abstract but is false for every configuration Rook can express: RGW's `validate_account_tenant` (enforced at user create and modify) requires an account member's tenant to equal the account's tenant, and `CephObjectStoreAccount` cannot create tenanted accounts — so any CR combining non-empty `tenant` with `accountRef` fails at RGW with `EINVAL` on a doubly-immutable field pair, permanently wedged until delete/recreate. Rejecting the combination at admission turns that into an immediate, clear error; the rule can be relaxed compatibly once accounts grow a tenant field.

````suggestion
## Interaction with `AccountRef`

RGW requires an account member's tenant to equal the account's tenant
(`validate_account_tenant`, enforced at user create and modify). Rook's
`CephObjectStoreAccount` cannot currently create tenanted accounts, so
every expressible `tenant` + `accountRef` combination would fail at RGW
with `EINVAL` on a doubly-immutable field pair, permanently wedging the
CR. The spec therefore rejects the combination at admission:

```go
// +kubebuilder:validation:XValidation:message="tenant cannot be combined with accountRef (CephObjectStoreAccount does not support tenants)",rule="!(has(self.tenant) && has(self.accountRef))"
```

Supporting tenanted account members later requires a tenant field on
`CephObjectStoreAccount` (go-ceph already transmits one on account
create/modify) and lifting this rule — a backward-compatible relaxation.
````

— the Overlord

Author: added this as future work scoped item 
```
  379 RGW requires an account member's tenant to equal the account's tenant                                                                    380 (`validate_account_tenant`, enforced at user create/modify). Rook's
  381 `CephObjectStoreAccount` cannot currently create tenanted accounts, so every                                                             382 expressible `tenant` + `accountRef` combination would fail at RGW with
  383 `EINVAL` on a doubly-immutable field pair. The spec therefore rejects the
  384 combination at admission (CEL rule above). Supporting tenanted account
  385 members requires adding a tenant field to `CephObjectStoreAccount` (go-ceph
  386 already transmits one) and is listed under [Future work](#future-work).
```
